funnel/models/video.py

```diff
@@ -0,0 +1,152 @@
+
+    @property
+    def video(self):
+        data = None
+        if self.video_source and self.video_id:
+            data = redis_store.hgetall(self.video_cache_key)
+            if data:
+                if 'uploaded_at' in data:
+                    data['uploaded_at'] = parse_isoformat(data['uploaded_at'])
+                if 'duration' in data:
+                    data['duration'] = int(data['duration'])
+            else:
+                data = {
+                    'source': self.video_source,
+                    'id': self.video_id,
+                    'url': self.video_url,
+                }
+                if self.video_source == 'youtube':
+                    youtube_video = requests.get(
+                        'https://www.googleapis.com/youtube/v3/videos?part=snippet,contentDetails&id={video_id}&key={api_key}'.format(
+                            video_id=self.video_id,
+                            api_key=app.config['YOUTUBE_API_KEY'],
+                        )
+                    ).json()
+                    if youtube_video is None or not youtube_video['items']:
+                        raise VideoException(
+                            "Unable to fetch data, please check the youtube url"
+                        )
+                    else:
+                        youtube_video = youtube_video['items'][0]
+
+                        data['duration'] = int(
+                            parse_duration(
+                                youtube_video['contentDetails']['duration']
+                            ).total_seconds()
+                        )
+                        data['uploaded_at'] = parse_isoformat(
+                            youtube_video['snippet']['publishedAt']
+                        ).isoformat()
+                        data['thumbnail'] = youtube_video['snippet']['thumbnails'][
+                            'medium'
+                        ]['url']
+                elif self.video_source == 'vimeo':
+                    vimeo_video = requests.get(
+                        "https://vimeo.com/api/v2/video/%s.json" % (self.video_id)
+                    ).json()
+                    if vimeo_video:
+                        vimeo_video = vimeo_video[0]
+
+                        data['duration'] = vimeo_video['duration']
+                        data['uploaded_at'] = parse_isoformat(
+                            vimeo_video['upload_date'], naive=True, delimiter=' '
+                        ).isoformat()
+                        data['thumbnail'] = vimeo_video['thumbnail_medium']
+                else:
+                    # source = raw
+                    data['duration'] = 0
+                    data['uploaded_at'] = None
+                    data['thumbnail'] = None
+                redis_store.hmset(self.video_cache_key, data)
+                redis_store.expire(
+                    self.video_cache_key, 60 * 60 * 24 * 2
+                )  # caching for 2 days
+        return data
```

Comment:
This is returning a string date. It should be cast back to datetime. This may be a good occasion to introduce JSON schemas into our code.